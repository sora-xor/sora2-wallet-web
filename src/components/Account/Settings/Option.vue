<template>
  <div :class="['settings-option', { disabled }]">
    <label :class="['settings-option-label', { disabled }]">
      <s-switch v-model="model" :disabled="disabled" />
      <div :class="['settings-option-label-description', { hint: withHint }]">
        <span class="settings-option-label-title">{{ title }}</span>
        <span v-if="withHint" class="settings-option-label-hint">{{ hint }}</span>
      </div>
    </label>
    <slot />
  </div>
</template>

<script lang="ts">
import { Component, ModelSync, Prop, Vue } from 'vue-property-decorator';

@Component
export default class AccountSettingsOption extends Vue {
  @Prop({ default: '', type: String }) readonly hint!: string;
  @Prop({ default: '', type: String }) readonly title!: string;
  @Prop({ default: false, type: Boolean }) readonly disabled!: boolean;
  @Prop({ default: false, type: Boolean }) readonly withHint!: boolean;

  @ModelSync('value', 'input', { type: Boolean })
  readonly model!: boolean;
}
</script>

<style lang="scss" scoped>
.settings-option {
  display: flex;
  flex-flow: column nowrap;
  gap: $basic-spacing-medium;
  width: 100%;

  &.disabled {
    opacity: 0.6;
  }
}

.settings-option-label {
  display: flex;
  flex-flow: row nowrap;
  align-items: center;
  gap: $basic-spacing-medium;

  &:not(.disabled) {
    cursor: pointer;
  }

  &-description {
    display: flex;
    flex-flow: column nowrap;
    align-items: flex-start;
    font-size: var(--s-font-size-medium);
    font-weight: 300;

    &.hint {
      font-size: var(--s-font-size-small);
    }
  }

  &-title {
    color: var(--s-color-base-content-primary);
  }

  &-hint {
    color: var(--s-color-base-content-secondary);
    font-size: var(--s-font-size-extra-small);
  }
}
</style>
