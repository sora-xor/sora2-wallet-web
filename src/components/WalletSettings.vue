<template>
  <wallet-base :title="t('settings.title')" show-back @back="handleBack">
    <div class="wallet-settings">
      <s-select :value="activeNetwork.id" :placeholder="t('settings.network')" @change="id => setActiveNetwork({ id })">
        <s-option
          v-for="network in availableNetworks"
          :key="network.id"
          :value="network.id"
          :label="network.name"
        />
      </s-select>
      <div
        v-for="(item, index) in menuTabs"
        :key="index"
        @click="navigate({ name: item.route })"
      >
        <div class="wallet-settings-item s-flex">
          <div class="wallet-settings-item-text s-flex">
            <span class="wallet-settings-item-text_main">{{ item.title }}</span>
            <span class="wallet-settings-item-text_secondary">{{ item.desc }}</span>
          </div>
          <s-icon class="wallet-settings-item_icon" name="chevron-right" :size="16" />
        </div>
        <s-divider v-if="index !== menuTabs.length - 1" class="wallet-settings-item_divider" />
      </div>
      <div class="wallet-settings_action s-flex">
        <s-button
          type="secondary"
          size="small"
          @click="onLogout"
        >
          {{ t('logoutText') }}
        </s-button>
      </div>
    </div>
  </wallet-base>
</template>

<script lang="ts">
import { Component, Mixins, Prop } from 'vue-property-decorator'
import { Action, Getter, State } from 'vuex-class'

import TranslationMixin from './mixins/TranslationMixin'
import WalletBase from './WalletBase.vue'
import { Network, RouteNames, SettingsMenu } from '../consts'

interface MenuItem {
  title: string;
  desc: string;
  route: RouteNames;
}

@Component({
  components: { WalletBase }
})
export default class WalletSettings extends Mixins(TranslationMixin) {
  readonly menuTabs: MenuItem[] = [
    {
      title: this.t(`settings.menu.${SettingsMenu.Language}.title`),
      desc: this.t(`settings.menu.${SettingsMenu.Language}.desc`),
      route: RouteNames.WalletSettingsLanguage
    },
    {
      title: this.t(`settings.menu.${SettingsMenu.Networks}.title`),
      desc: this.t(`settings.menu.${SettingsMenu.Networks}.desc`),
      route: RouteNames.WalletSettingsNetworks
    },
    {
      title: this.t(`settings.menu.${SettingsMenu.About}.title`),
      desc: this.t(`settings.menu.${SettingsMenu.About}.desc`),
      route: RouteNames.WalletSettingsAbout
    }
  ]

  @State(state => state.Settings.activeNetwork) activeNetwork!: Network

  @Getter availableNetworks!: Array<Network>

  @Action navigate
  @Action logout
  @Action setActiveNetwork

  handleBack (): void {
    this.navigate({ name: RouteNames.Wallet })
  }

  onLogout (): void {
    this.navigate({ name: RouteNames.WalletConnection })
    this.logout()
  }
}
</script>

<style scoped lang="scss">
.wallet-settings {
  &_action {
    justify-content: center;
    > * {
      width: 100%;
    }
  }
  &-item {
    justify-content: space-between;
    padding: $basic-spacing_small 0;
    &-text {
      flex-direction: column;
      &_main {
        font-size: $font-size_normal;
        line-height: 1.8;
      }
      &_secondary {
        font-size: $font-size_small;
        color: var(--s-color-base-content-tertiary);
        line-height: 1.8;
      }
    }
    &_icon {
      margin: auto 0;
    }
    &_divider {
      margin: unset;
    }
    &:hover {
      cursor: pointer;
      .wallet-settings-item-text_main {
        color: var(--s-color-button-tertiary-color);
      }
    }
  }
}
</style>
