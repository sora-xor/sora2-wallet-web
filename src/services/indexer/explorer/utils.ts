import { formatStringNumber } from '../../../util';

import type { SubqueryAssetEntity } from '../subquery/types';
import type { SubsquidAssetEntity } from '../subsquid/types';
import type { FiatPriceObject, UpdatesStream, PoolApyObject } from '../types';

export function parseAssetFiatPrice(entity: SubsquidAssetEntity | SubqueryAssetEntity): FiatPriceObject {
  const acc = {};
  const id = entity.id;
  const priceFPNumber = formatStringNumber(entity.priceUSD);
  const isPriceFinity = priceFPNumber.isFinity();
  if (isPriceFinity) {
    acc[id] = priceFPNumber.toCodecString();
  }
  return acc;
}

export function parsePriceStreamUpdate(entity: UpdatesStream): Nullable<FiatPriceObject> {
  if (!entity?.data) return null;

  const data = JSON.parse(entity.data);

  return Object.entries(data).reduce((acc, [id, price]) => {
    const priceFPNumber = formatStringNumber(price as string);
    const isPriceFinity = priceFPNumber.isFinity();
    if (isPriceFinity) {
      acc[id] = priceFPNumber.toCodecString();
    }
    return acc;
  }, {});
}

export function parseApyStreamUpdate(entity: UpdatesStream): PoolApyObject {
  const data = entity?.data ? JSON.parse(entity.data) : {};

  return Object.entries(data).reduce((acc, [id, apy]) => {
    const strategicBonusApyFPNumber = formatStringNumber(apy as string);
    const isStrategicBonusApyFinity = strategicBonusApyFPNumber.isFinity();
    if (isStrategicBonusApyFinity) {
      acc[id] = strategicBonusApyFPNumber.toCodecString();
    }
    return acc;
  }, {});
}
