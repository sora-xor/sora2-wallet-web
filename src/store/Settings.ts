import map from 'lodash/fp/map'
import flatMap from 'lodash/fp/flatMap'
import fromPairs from 'lodash/fp/fromPairs'
import flow from 'lodash/fp/flow'
import concat from 'lodash/fp/concat'

import { storage } from '../util/storage'
import i18n from '../lang'
import { Network, WalletPermissions } from '../consts'

const types = flow(
  flatMap(x => [x + '_REQUEST', x + '_SUCCESS', x + '_FAILURE']),
  concat([
    'SET_LOCALE',
    'ADD_NETWORK',
    'SET_ACTIVE_NETWORK',
    'SET_PERMISSIONS'
  ]),
  map(x => [x, x]),
  fromPairs
)([])

export const DEFAULT_NETWORKS: Array<Network> = [{
  id: 1,
  name: 'SORA-staging Testnet',
  address: 'wss://ws.stage.sora2.soramitsu.co.jp',
  explorer: 'https://polkascan.io',
  editable: false
}]

function modifyActiveNetwork (state, network: Network) {
  // For now it works as remove operation
  // If it should be removed then DEFAULT_NETWORKS[0] will be set as active network
  if (network && network.name && network.address && network.explorer) {
    state.activeNetwork = network
    storage.set('activeNetwork', JSON.stringify(network))
  } else {
    state.activeNetwork = DEFAULT_NETWORKS[0]
    storage.set('activeNetwork', JSON.stringify(DEFAULT_NETWORKS[0]))
  }
}

function initialState () {
  return {
    locale: storage.get('locale') || i18n.locale,
    locales: i18n.availableLocales,
    activeNetwork: (JSON.parse(storage.get('activeNetwork') as string) || DEFAULT_NETWORKS[0]) as Network,
    customNetworks: (JSON.parse(storage.get('networks') as string) || []) as Array<Network>,
    networks: DEFAULT_NETWORKS,
    permissions: {
      sendAssets: true,
      swapAssets: true
    }
  }
}

const state = initialState()

const getters = {
  availableNetworks (state) {
    return [...state.networks, ...state.customNetworks]
  },
  activeNetwork (state) {
    return state.activeNetwork
  },
  permissions (state) {
    return state.permissions
  }
}

const mutations = {
  [types.RESET] (state) {
    const s = initialState()
    Object.keys(s).forEach(key => {
      state[key] = s[key]
    })
  },

  [types.SET_LOCALE] (state, { locale }) {
    i18n.locale = locale
    state.locale = locale
    storage.set('locale', locale)
  },

  [types.ADD_NETWORK] (state, { network }) {
    const alreadyExistingNetwork = state.customNetworks.find(({ id }) => id === network.id) as Network
    if (alreadyExistingNetwork) {
      alreadyExistingNetwork.name = network.name
      alreadyExistingNetwork.address = network.address
      alreadyExistingNetwork.explorer = network.explorer
      if (state.activeNetwork.id === alreadyExistingNetwork.id) {
        modifyActiveNetwork(state, alreadyExistingNetwork)
      }
    }
    const customNetworks = (alreadyExistingNetwork ? state.customNetworks : [...state.customNetworks, network])
      // For now it works as remove operation
      .filter(({ name, address, explorer }) => name && address && explorer)
    storage.set('networks', JSON.stringify(customNetworks))
    state.customNetworks = customNetworks
  },

  [types.SET_ACTIVE_NETWORK] (state, { network }) {
    storage.set('activeNetwork', JSON.stringify(network))
    state.activeNetwork = network
    // TODO: add connection
  },

  [types.SET_PERMISSIONS] (state, permissions: WalletPermissions) {
    if (typeof permissions !== 'object' || Array.isArray(permissions)) {
      console.error(`Permissions should be an object, ${typeof permissions} is given`)
      return
    }
    state.permissions = { ...state.permissions, ...permissions }
  }
}

const actions = {
  setLocale ({ commit }, { locale }) {
    commit(types.SET_LOCALE, { locale })
  },

  addNetwork ({ commit }, { network }) {
    commit(types.ADD_NETWORK, { network })
  },

  setActiveNetwork ({ commit, getters }, { id }) {
    const network = getters.availableNetworks.find(net => net.id === id)
    commit(types.SET_ACTIVE_NETWORK, { network })
  },

  setPermissions ({ commit }, permisssions) {
    commit(types.SET_PERMISSIONS, permisssions)
  }
}

export default {
  types,
  state,
  getters,
  mutations,
  actions
}
