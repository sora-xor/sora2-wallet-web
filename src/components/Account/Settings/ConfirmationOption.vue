<template>
  <account-settings-option
    :title="t('accountSettings.confirmation.title')"
    :hint="t('accountSettings.hint')"
    :with-hint="withHint"
    v-model="model"
  >
    <slot />
  </account-settings-option>
</template>

<script lang="ts">
import { Component, Mixins, Prop } from 'vue-property-decorator';

import { mutation, state } from '../../../store/decorators';
import TranslationMixin from '../../mixins/TranslationMixin';

import AccountSettingsOption from './Option.vue';

@Component({
  components: {
    AccountSettingsOption,
  },
})
export default class AccountConfirmationOption extends Mixins(TranslationMixin) {
  @Prop({ default: false, type: Boolean }) readonly withHint!: boolean;

  @state.transactions.isConfirmTxDialogDisabled private isConfirmTxDialogDisabled!: boolean;
  @mutation.transactions.setConfirmTxDialogDisabled private setConfirmTxDialogDisabled!: (flag: boolean) => void;

  get model(): boolean {
    return this.isConfirmTxDialogDisabled;
  }

  set model(value: boolean) {
    this.setConfirmTxDialogDisabled(value);
  }
}
</script>
