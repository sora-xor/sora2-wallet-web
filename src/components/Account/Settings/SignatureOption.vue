<template>
  <account-settings-option
    :disabled="disabled"
    :title="t('accountSettings.signature.title')"
    :hint="t('accountSettings.hint')"
    :with-hint="withHint"
    v-model="model"
  >
    <slot />
    <div v-if="disabled || model" class="save-password-duration">
      <template v-if="passwordResetDate">
        <div class="save-password-duration-saved">
          <span class="save-password-duration-title">{{ t('accountSettings.disabled') }}:</span>
          <span> {{ passwordResetDate }}</span>
        </div>
      </template>
      <template v-else>
        <span class="save-password-duration-title">{{ t('accountSettings.disable') }}:</span>
        <s-tabs v-model="passwordTimeoutModel" type="rounded" class="save-password-durations">
          <s-tab
            v-for="duration in durations"
            :key="duration"
            :label="duration"
            :name="duration"
            :disabled="disabled"
          />
        </s-tabs>
      </template>
    </div>
  </account-settings-option>
</template>

<script lang="ts">
import dayjs from 'dayjs';
import relativeTime from 'dayjs/plugin/relativeTime';
import { Component, Mixins, Prop } from 'vue-property-decorator';

import { PassphraseTimeout, PassphraseTimeoutDuration, DefaultPassphraseTimeout } from '../../../consts';
import { action, mutation, state } from '../../../store/decorators';
import TranslationMixin from '../../mixins/TranslationMixin';

import AccountSettingsOption from './Option.vue';

dayjs.extend(relativeTime);

@Component({
  components: {
    AccountSettingsOption,
  },
})
export default class AccountSignatureOption extends Mixins(TranslationMixin) {
  @Prop({ default: false, type: Boolean }) readonly withHint!: boolean;
  @Prop({ default: false, type: Boolean }) readonly disabled!: boolean;

  @state.transactions.isSignTxDialogDisabled private isSignTxDialogDisabled!: boolean;
  @mutation.transactions.setSignTxDialogDisabled private setSignTxDialogDisabled!: (flag: boolean) => void;

  @state.account.accountPasswordTimeout private accountPasswordTimeout!: number;
  @mutation.account.setPasswordTimeout private setPasswordTimeout!: (timeout: number) => void;

  @state.account.address private connected!: string;
  @state.account.accountPasswordTimestamp private accountPasswordTimestamp!: Record<string, Nullable<number>>;
  @action.account.resetAccountPassphrase private resetAccountPassphrase!: (address: string) => void;

  readonly durations = PassphraseTimeout;

  get model(): boolean {
    return this.isSignTxDialogDisabled;
  }

  set model(value: boolean) {
    this.setSignTxDialogDisabled(value);

    if (!value) {
      this.resetAccountPassphrase(this.connected);
    }
  }

  get passwordTimeoutModel(): PassphraseTimeout {
    const key = Object.keys(PassphraseTimeoutDuration).find(
      (key) => PassphraseTimeoutDuration[key] === this.accountPasswordTimeout
    );

    if (!key) return PassphraseTimeout.FIFTEEN_MINUTES;

    return key as PassphraseTimeout;
  }

  set passwordTimeoutModel(name: PassphraseTimeout) {
    const duration = PassphraseTimeoutDuration[name] ?? DefaultPassphraseTimeout;
    this.setPasswordTimeout(duration);
  }

  private timestamp: Nullable<number> = null;
  private timer: Nullable<NodeJS.Timeout> = null;

  created(): void {
    this.createTimer();
  }

  destroyed(): void {
    this.resetTimer();
  }

  private updateTimestamp(): void {
    this.timestamp = Date.now();
  }

  private createTimer(): void {
    this.resetTimer();
    this.updateTimestamp();
    this.timer = setInterval(this.updateTimestamp, 1000);
  }

  private resetTimer(): void {
    if (this.timer) {
      clearInterval(this.timer);
    }
    this.timer = null;
    this.timestamp = null;
  }

  get passwordResetDate(): Nullable<string> {
    const accountPasswordTimestamp = this.accountPasswordTimestamp[this.connected];

    if (!(accountPasswordTimestamp && this.timestamp)) return null;

    const diff = accountPasswordTimestamp + this.accountPasswordTimeout - this.timestamp;

    return dayjs.duration(diff).locale(this.dayjsLocale).humanize();
  }
}
</script>

<style lang="scss">
$telegram-web-app-width: 500px;

.save-password-durations {
  .el-tabs__header {
    margin-bottom: 0;
  }

  &.s-tabs.s-rounded .el-tabs__nav-wrap .el-tabs__item {
    text-transform: initial;

    &.is-disabled {
      cursor: not-allowed;
    }
  }

  // override s-tabs arrows
  .el-icon-arrow-left,
  .el-icon-arrow-right {
    display: block;
    margin-top: 10px;
  }

  &.s-tabs.s-rounded .el-tabs__header {
    @media screen and (max-width: $telegram-web-app-width) {
      display: flex;
      justify-content: center;
      align-items: center;
      width: auto;

      .el-tabs__nav-wrap .el-tabs__item.is-active {
        box-shadow: 0 0 2px 1px var(--s-shadow-color-dark);
      }
    }
  }

  .el-tabs__nav-scroll {
    @media screen and (max-width: $telegram-web-app-width) {
      width: 260px;
    }
  }
}
</style>

<style lang="scss" scoped>
.save-password-duration {
  display: flex;
  flex-flow: column nowrap;
  gap: $basic-spacing-small;

  &-title {
    font-size: var(--s-font-size-extra-small);
    font-weight: 700;
    text-transform: uppercase;
  }

  &-saved {
    display: flex;
    flex-flow: row wrap;
    align-items: baseline;
    gap: $basic-spacing;
  }
}
</style>
