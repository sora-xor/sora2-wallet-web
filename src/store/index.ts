import Vue from 'vue';
import Vuex from 'vuex';

import Account from './Account';
import Router from './Router';
import Settings from './Settings';
import Transactions from './Transactions';

const modules = {
  Account,
  Router,
  Settings,
  Transactions,
};

Vue.use(Vuex);

const store = new Vuex.Store({
  modules,
  strict: false,
});

export { modules };

export default store;
