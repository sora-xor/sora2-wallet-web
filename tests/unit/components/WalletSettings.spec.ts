import Vuex from 'vuex'
import { shallowMount, createLocalVue } from '@vue/test-utils'

import WalletSettings from '@/components/WalletSettings.vue'
import { DEFAULT_NETWORKS } from '@/store/Settings'
import { TranslationMock, SoramitsuElementsImport } from '../../utils'

const localVue = createLocalVue()
localVue.use(Vuex)
const store = new Vuex.Store({
  modules: {
    Router: {
      actions: {
        navigate: jest.fn()
      }
    },
    Account: {
      actions: {
        logout: jest.fn()
      }
    },
    Settings: {
      state: {
        activeNetwork: DEFAULT_NETWORKS[0]
      },
      getters: {
        availableNetworks: () => ([
          {
            id: 1,
            name: 'Main Ethereum Network',
            address: 'https://api.infura.io/v1/jsonrpc/mainnet',
            explorer: 'https://etherscan.io',
            editable: false
          }
        ])
      },
      actions: {
        setActiveNetwork: jest.fn()
      }
    }
  } as any
})

describe('WalletSettings.vue', () => {
  beforeEach(() => {
    SoramitsuElementsImport(localVue)
    TranslationMock(WalletSettings)
  })

  it('should renders correctly', () => {
    const wrapper = shallowMount(WalletSettings, { localVue, store })
    expect(wrapper.element).toMatchSnapshot()
  })
})
