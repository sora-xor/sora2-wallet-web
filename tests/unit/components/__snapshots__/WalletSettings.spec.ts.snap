// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletSettings.vue should renders correctly 1`] = `
<wallet-base-stub
  showback="true"
  title=""
>
  <div
    class="wallet-settings"
  >
    <s-select-stub
      autocomplete="off"
      borderradius="mini"
      disabled="true"
      inputtype="input"
      loadingtext="Loading"
      multiplelimit="0"
      multipletextprefix=""
      name=""
      nodatatext="No data"
      placeholder="Select"
      popperclass=""
      value="1"
    >
      <s-option-stub
        label="Main Ethereum Network"
        value="1"
      />
    </s-select-stub>
     
    <div>
      <div
        class="wallet-settings-item s-flex"
      >
        <div
          class="wallet-settings-item-text s-flex"
        >
          <span
            class="wallet-settings-item-text_main"
          >
            
          </span>
           
          <span
            class="wallet-settings-item-text_secondary"
          >
            
          </span>
        </div>
         
        <s-icon-stub
          class="wallet-settings-item_icon"
          name="chevron-right"
          size="12px"
        />
      </div>
       
      <s-divider-stub
        class="wallet-settings-item_divider"
        contentposition="center"
        direction="horizontal"
        type="secondary"
      />
    </div>
    <div>
      <div
        class="wallet-settings-item s-flex"
      >
        <div
          class="wallet-settings-item-text s-flex"
        >
          <span
            class="wallet-settings-item-text_main"
          >
            
          </span>
           
          <span
            class="wallet-settings-item-text_secondary"
          >
            
          </span>
        </div>
         
        <s-icon-stub
          class="wallet-settings-item_icon"
          name="chevron-right"
          size="12px"
        />
      </div>
       
      <s-divider-stub
        class="wallet-settings-item_divider"
        contentposition="center"
        direction="horizontal"
        type="secondary"
      />
    </div>
    <div>
      <div
        class="wallet-settings-item s-flex"
      >
        <div
          class="wallet-settings-item-text s-flex"
        >
          <span
            class="wallet-settings-item-text_main"
          >
            
          </span>
           
          <span
            class="wallet-settings-item-text_secondary"
          >
            
          </span>
        </div>
         
        <s-icon-stub
          class="wallet-settings-item_icon"
          name="chevron-right"
          size="12px"
        />
      </div>
       
      <!---->
    </div>
     
    <div
      class="wallet-settings_action s-flex"
    >
      <s-button-stub
        borderradius="small"
        icon=""
        iconposition="left"
        nativetype="button"
        size="small"
        tabindex="0"
        tooltip=""
        tooltipplacement="top"
        type="secondary"
      >
        
        
      
      </s-button-stub>
    </div>
  </div>
</wallet-base-stub>
`;
